src/lib/compression/src/lib.rs

```diff
@@ -1,45 +1,45 @@
 #![no_std]
 
-use core::{ptr::read_volatile, slice};
+use core::slice;
 use log::println;
 
-// offset_bits aka EI, usually 10..13
-// length_bits aka EJ, usually 4..5
-// buf_fill_byte (often 0x20, space) - we use 0x00, the zero byte
-// decompress_buf_size, always 1 << EI
-// compress_buf_size, always 2 << EI
-const EI: usize = 12;
-const DECOMP_BUF_SIZE: usize = 1 << EI;
-const COMP_BUF_SIZE: usize = 2 << EI;
-pub type OreLzss = lzss::Lzss<EI, 4, 0x00, DECOMP_BUF_SIZE, COMP_BUF_SIZE>;
+use miniz_oxide::inflate as moi;
+use miniz_oxide::inflate::core as moc;
+
+// This flag means inflate should try parsing a zlib header.
+// We provide zlib compressed payloads including the header by convetion.
+const FLAGS: u32 = moc::inflate_flags::TINFL_FLAG_PARSE_ZLIB_HEADER;
 
 /// # Safety
 ///
 /// The caller needs to ensure correctness of source and target addresses.
-/// The first four bytes at source address must hold the compressed size.
 /// The payload size is the maximum expected size after decompression.
 /// Ensure that it fits, i.e., `payload_size` bytes after target address
 /// are unused. We allocate a buffer of that size; exceeding it will error.
-pub unsafe fn decompress(source_addr: usize, target_addr: usize, payload_size: usize) {
-    // NOTE: This must be u32 because it is 4 bytes, so cast later.
-    let compressed_size = unsafe { read_volatile(source_addr as *const u32) };
-    let in_ptr = (source_addr + 4) as *const u8;
+pub unsafe fn decompress(
+    source_addr: usize,
+    target_addr: usize,
+    compressed_size: usize,
+    payload_size: usize,
+) {
+    let in_ptr = source_addr as *const u8;
     let out_ptr = target_addr as *mut u8;
-    let input = unsafe { slice::from_raw_parts(in_ptr, compressed_size as usize) };
+    let input = unsafe { slice::from_raw_parts(in_ptr, compressed_size) };
     let output = unsafe { slice::from_raw_parts_mut(out_ptr, payload_size) };
```

Comment:
This does not seem save to me; taking a slice of uninit memory is UB. If you memset it to zero first, that's ok. But I'd say it's even better for this function to just take a mutable slice of bytes to expand into.

Reply:
It's not that it may mess with the allocator per se, but rather that the behavior is undefined because it violates the language's rules.